src/solc_0.8/defi/rules/ContributionRules.sol

```diff
@@ -0,0 +1,194 @@
+        require(_islistERC721Member(reqContract), "ContributionRules: contract is not in the list");
+        uint256 indexToDelete = _listERC721[reqContract].index;
+        IERC721 addrToMove = _listERC721Index[_listERC721Index.length - 1];
+        _listERC721Index[indexToDelete] = addrToMove;
+        _listERC721[addrToMove].index = indexToDelete;
+        _listERC721Index.pop();
+    }
+
+    function delMultiplierERC1155fromList(IERC1155 reqContract) external onlyOwner {
+        require(reqContract != IERC1155(address(0)), "ContributionRules: invalid contract address");
+        require(_islistERC1155Member(reqContract), "ContributionRules: contract is not in the list");
+        uint256 indexToDelete = _listERC1155[reqContract].index;
+        IERC1155 addrToMove = _listERC1155Index[_listERC1155Index.length - 1];
+        _listERC1155Index[indexToDelete] = addrToMove;
+        _listERC1155[addrToMove].index = indexToDelete;
+        _listERC1155Index.pop();
+    }
+
+    function _islistERC721Member(IERC721 reqContract) internal view returns (bool) {
+        if (_listERC721Index.length == 0) return false;
+        return (_listERC721Index[_listERC721[reqContract].index] == reqContract);
+    }
+
+    function _islistERC1155Member(IERC1155 reqContract) internal view returns (bool) {
+        if (_listERC721Index.length == 0) return false;
+        return (_listERC1155Index[_listERC1155[reqContract].index] == reqContract);
+    }
+
+    function _multiplierBalanceOfLand(address account) internal view returns (uint256) {
+        uint256 numLands = landMultiplierContract.balanceOf(account);
+
+        if (numLands == 0) {
+            return 0;
+        }
+
+        uint256 _landMultiplier =
+            NFT_FACTOR_6 * (NFT_CONSTANT_3 + SafeMathWithRequire.cbrt3((((numLands - 1) * ROOT3_FACTOR) + 1)));
+        if (_landMultiplier > MIDPOINT_9) {
+            _landMultiplier = MIDPOINT_9 + (_landMultiplier - MIDPOINT_9) / 10;
+        }
+
+        return _landMultiplier / DECIMALS_7;
+    }
+
+    function _multiplierBalanceOfERC721(address account) internal view returns (uint256) {
+        uint256 _multiplier = 0;
+        uint256 _bal = 0;
+
+        for (uint256 i = 0; i < _listERC721Index.length; i++) {
+            IERC721 reqContract = _listERC721Index[i];
+
+            if (_listERC721[reqContract].balanceOf == true) {
+                _bal = reqContract.balanceOf(account);
+                _multiplier = _multiplier + (_bal * _listERC721[reqContract].multiplier[i]);
+            } else {
+                for (uint256 j = 0; j < _listERC721[reqContract].ids.length; j++) {
+                    address owner = reqContract.ownerOf(_listERC721[reqContract].ids[j]);
+                    if (owner == account) {
+                        ++_bal;
+                    }
+                    _multiplier = _multiplier + (_bal * _listERC721[reqContract].multiplier[j]);
+                }
+            }
+        }
```

Comment:
split the function in 2
(the 2nd loop for instance)

Reply:
not sure that this is really necessary